qwerty_ast/src/typecheck.rs

```diff
@@ -0,0 +1,456 @@
+                    },
+                    span: span.clone(),
+                });
+            }
+            Ok(())
+        }
+    }
+}
+
+//
+// ─── EXPRESSIONS ────────────────────────────────────────────────────────────────
+//
+
+/// Typecheck an expression and return its type.
+pub fn typecheck_expr(expr: &Expr, env: &mut TypeEnv) -> Result<Type, TypeError> {
+    match expr {
+        Expr::Variable { name, span } => {
+            env.get_var(name)
+                .cloned()
+                .ok_or(TypeError {
+                    kind: TypeErrorKind::UndefinedVariable(name.clone()),
+                    span: span.clone(),
+                })
+        }
+
+        Expr::UnitLiteral { span: _ } => Ok(Type::UnitType),
+
+        Expr::Adjoint { func, span: _ } => {
+            // Adjoint should be a function type (unitary/quantum), not classical.
+            let func_ty = typecheck_expr(func, env)?;
+            // TODO: Enforce Qwerty adjoint typing rules.
+            Ok(func_ty)
+        }
+
+        Expr::Pipe { lhs, rhs, span: _ } => {
+            // Typing rule: lhs type must match rhs function input type.
+            let lhs_ty = typecheck_expr(lhs, env)?;
+            let rhs_ty = typecheck_expr(rhs, env)?;
+            match &rhs_ty {
+                Type::FuncType { in_ty, out_ty } => {
+                    if **in_ty != lhs_ty {
+                        return Err(TypeError {
+                            kind: TypeErrorKind::MismatchedTypes {
+                                expected: format!("{:?}", in_ty),
+                                found: format!("{:?}", lhs_ty),
+                            },
+                            span: None,
+                        });
+                    }
+                    Ok((**out_ty).clone())
+                }
+                _ => Err(TypeError {
+                    kind: TypeErrorKind::NotCallable(format!("{:?}", rhs_ty)),
+                    span: None,
+                }),
+            }
+        }
+
+        Expr::Measure { basis, span: _ } => {
+            // Qwerty: measurement returns classical result; basis must be valid.
+            typecheck_basis(basis, env)?; //  is it a legal quantum basis?
+
+            Ok(Type::RegType {
+                elem_ty: RegKind::Bit,
+                dim: 1, // TODO: Make dynamic based on basis (Check with Austin about its validation)
+                // Self Note (verify with Austin): Currently this measurement returns a single classical bit
+                // But in real quantum programs, we might measure multiple qubits at once (e.g measuring a register of 3 qubits gives you 3 classical bits, ryt?)
+                // The number of bits returned should depend on the size/dimension of the basis being measured
+                // So, make 'dim' reflect the actual number of qubits measured, as determined by the basis argument
+                // Need to understand and work on dimensions.. Discuss bro!
```

Comment:
Yeah exactly dude, `dim` here should be the dimension (number of qubits) of the basis